packages/x-data-grid-pro/src/hooks/features/serverSideData/utils.ts

```diff
@@ -0,0 +1,120 @@
+import { GridRowId } from '@mui/x-data-grid';
+import { GridPrivateApiPro } from '../../../models/gridApiPro';
+
+// Make these configurable using dedicated props?
+const MAX_CONCURRENT_REQUESTS = Infinity;
+const QUEUE_PROCESS_INTERVAL_MS = 300;
+
+export const runIfServerMode = (modeProp: 'server' | 'client', fn: Function) => () => {
+  if (modeProp === 'server') {
+    fn();
+  }
+};
+
+export enum RequestStatus {
+  INQUEUE,
+  PENDING,
+  SETTLED,
+  UNKNOWN,
+}
+
+/**
+ * Fetches row children from the server with option to limit the number of concurrent requests
+ * Determines the status of a request based on the enum `RequestStatus`
+ * Uses `GridRowId` to uniquely identify a request
+ */
+export class NestedDataManager {
+  private pendingRequests: Set<GridRowId> = new Set();
+
+  private queuedRequests: Set<GridRowId> = new Set();
+
+  private settledRequests: Set<GridRowId> = new Set();
+
+  private api: GridPrivateApiPro;
+
+  private maxConcurrentRequests: number;
+
+  private queueProcessInterval: number;
+
+  private timer?: string | number | NodeJS.Timeout;
+
+  constructor(
+    privateApiRef: React.MutableRefObject<GridPrivateApiPro>,
+    maxConcurrentRequests = MAX_CONCURRENT_REQUESTS,
+    queueProcessInterval = QUEUE_PROCESS_INTERVAL_MS,
+  ) {
+    this.api = privateApiRef.current;
+    this.maxConcurrentRequests = maxConcurrentRequests;
+    this.queueProcessInterval = queueProcessInterval;
+  }
+
+  private processQueue = async () => {
+    if (this.queuedRequests.size === 0) {
+      clearInterval(this.timer);
+      return;
+    }
+    if (this.pendingRequests.size >= this.maxConcurrentRequests) {
+      return;
+    }
+    const fetchQueue = Array.from(this.queuedRequests);
+    for (let i = 0; i < this.maxConcurrentRequests; i += 1) {
+      const nextId = fetchQueue[i];
+      if (!nextId) {
+        clearInterval(this.timer);
+        return;
+      }
+      this.queuedRequests.delete(nextId);
```

Comment:
Should be run before the `nextId` check?

Reply:
It removes the value from `queuedRequests`, the `nextId` check ensures it's present in the `queuedRequests`, if it's not present, it doesn't make sense to remove it, right?

Reply:
If we assume `queuedRequests` can contain `undefined` as a value then we should remove it from `queuedRequests`

A better option is to ensure we don't put `undefined` there. In that case, we won't need the `typeof nextId === 'undefined'` check 🙂

Reply:
I refactored this code a bit, now we don't need `typeof nextId === 'undefined'` check as you suggested.